### PR TITLE
Add async middleware; closes #14

### LINKED ISSUE #14: Async support in Django 3.0+
Django 3.0+ is beginning to provide async support: https://docs.djangoproject.com/en/3.1/topics/async/

We will be working on an async interface to our underlying Python library soon: https://github.com/ipinfo/python/issues/25

When it's more clear how Django plans on supporting a fully async request stack, we can utilize the async interface to the Python SDK.

### PR DESCRIPTION
**Pull request by valentinogagliardi**
Hi! First of all thanks for this package! As I was integrating this middleware in a project of mine, I noticed it lacked async support. This PR aims at adding an asynchronous Django middleware.

### CODE REVIEW

**Review by UmanShahzad — COMMENTED** [review]
I strongly appreciate this contribution! Please lemme know when it's ready and I'll review.

**Comment by valentinogagliardi** [comment]
Thanks, I think this is ready. The only problem I'm seeing is that in the test for Django 3.1, the async client does not sets `HTTP_USER_AGENT`.

**Comment by UmanShahzad** [comment]
Alrighty, thanks a lot! I'll try my best to check this out this upcoming week.

**Comment by valentinogagliardi** [comment]
@UmanShahzad there are still a couple of things to take into account here that popped up while using this middleware in production:

- how does `ipinfo.getDetails()` determines the ip to query for?
- when the Django project is running under a reverse proxy, the middleware incorrectly queries for the ip of the reverse proxy, not for the ip defined in `"HTTP_X_FORWARDED_FOR"`

I still need to handle the latter case.

**Comment by UmanShahzad** [comment]
@valentinogagliardi Yes this is an issue related to the IP selection process - solving https://github.com/ipinfo/django/issues/15 allows us to modify it, to e.g. use an IP from a reverse proxy, or some totally other custom user strategy.

However you're correct in that the current strategy of just doing a raw `ipinfo.getDetails` makes no sense - this is just going to get the IP of the server everytime. We should be passing it the IP on the request, or the IP after running the selection handler that'd be made from solving https://github.com/ipinfo/django/issues/15.

**Comment by valentinogagliardi** [comment]
@UmanShahzad that makes sense. At this point, at least in my fork, I can explicitly pass the argument to `getDetails()` to make this usable

File: setup.py [hunk 1]
```diff
@@ -10,14 +10,14 @@
...
+    version="1.2.0",
     description="Official Django library for IPinfo",
     long_description=long_description,
     url="https://github.com/ipinfo/django",
     author="IPinfo",
     author_email="support@ipinfo.io",
     license="Apache License 2.0",
     packages=["ipinfo_django"],
-    install_requires=["django", "ipinfo==3.0.0"],
+    install_requires=["django", "ipinfo==4.0.0"],
```
Comment (UmanShahzad) [inline setup.py:21]: We can upgrade to 4.2.0.

File: ipinfo_django/middleware.py [hunk 2]
```diff
@@ -32,8 +34,39 @@ def process_request(self, request):
...
+    def __call__(self, request):
+        return self.__acall__(request)
+
+    async def __acall__(self, request):
+        """Middleware hook that acts on and modifies request object."""
+        try:
+            if self.filter and self.filter(request):
+                request.ipinfo = None
+            else:
+                request.ipinfo = await self.ipinfo.getDetails(get_ip(request))
```
Comment (UmanShahzad) [inline ipinfo_django/middleware.py:63]: Could we also have the sync handler do this `get_ip` part?

File: requirements/compile.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+from pathlib import Path
+
+if __name__ == "__main__":
+    os.chdir(Path(__file__).parent)
+    os.environ["CUSTOM_COMPILE_COMMAND"] = "requirements/compile.py"
+    common_args = ["-m", "piptools", "compile", "--generate-hashes"] + sys.argv[1:]
+
+    subprocess.run(
+        [
+            "python3.9",
```
Comment (UmanShahzad) [inline requirements/compile.py:13]: I noticed we're tying ourselves to 3.9 a lot - any particular reason?

Comment (valentinogagliardi) [reply requirements/compile.py:13]: No particular reason. We can also add Python 3.10.

File: ipinfo_django/middleware.py (hunk 2 above)
Comment (valentinogagliardi) [reply ipinfo_django/middleware.py:63]: Do you mean here https://github.com/valentinogagliardi/django/blob/feature/async-middleware/ipinfo_django/middleware.py#L31?

Comment (UmanShahzad) [reply ipinfo_django/middleware.py:63]: Yep!

File: requirements/compile.py (hunk 3 above)
Comment (UmanShahzad) [reply requirements/compile.py:13]: No worries, all good for now.

**Comment by UmanShahzad** [comment]
I'll give it a quick test then merge, great work!

**Pull request merged**